### ISSUE TITLE
BUG: strings 'nan' istead of NaN values if pandas DataFrame is enlarged by setting values using DataFrame.at[] or DataFrame.loc[] with non-existent indexers

### ISSUE THREAD

**Issue by EBenkler** [pandas-dev/pandas#56379]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
df= pd.DataFrame() # empty dataframe to append configuration strings
df.at['A','1']="first string"
df.at['B','2']="second string"
print(type(df.iat[1,0]))
print(type(df.iat[0,1]))
# expected output:
# <class 'float'>
# <class 'float'>
# actual output:
# <class 'float'>
# <class 'numpy.str_'>
```

### Issue Description

When enlarging a DataFrame by setting values using DataFrame.at[] with non-existing indexers (see, e.g. [Setting with enlargement](https://pandas.pydata.org/docs/user_guide/indexing.html#setting-with-enlargement)), values at "unset" indices should be set to NaN, which has type <class 'float'>. This used to work in older versions of Pandas, but now, the nan-values in the first row are set to the string 'nan' with type <class 'numpy.str_'>. See the minimal working example, which shows the same behaviour if at[] is substituted by loc[].

### Expected Behavior

The above MWE should output:
<class 'float'>
<class 'float'>
Instead, it outputs:
<class 'float'>
<class 'numpy.str_'>
This is because the nan value in 
df.iat[0,1] or df.at['A','2'] is a string 'nan' instead of float nan.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit              : 2a953cf80b77e4348bf50ed724f8abc0d814d9dd
python              : 3.11.6.final.0
python-bits         : 64
OS                  : Windows
OS-release          : 10
Version             : 10.0.19045
machine             : AMD64
processor           : Intel64 Family 6 Model 42 Stepping 7, GenuineIntel
byteorder           : little
LC_ALL              : None
LANG                : en
LOCALE              : de_DE.cp1252

pandas              : 2.1.3
numpy               : 1.26.0
pytz                : 2023.3.post1
dateutil            : 2.8.2
setuptools          : 68.2.2
pip                 : 23.3.1
Cython              : None
pytest              : None
hypothesis          : None
sphinx              : 7.2.6
blosc               : None
feather             : None
xlsxwriter          : None
lxml.etree          : None
html5lib            : None
pymysql             : None
psycopg2            : None
jinja2              : 3.1.2
IPython             : 8.17.2
pandas_datareader   : None
bs4                 : 4.12.2
bottleneck          : None
dataframe-api-compat: None
fastparquet         : None
fsspec              : 2023.10.0
gcsfs               : None
matplotlib          : 3.8.1
numba               : None
numexpr             : 2.8.7
odfpy               : None
openpyxl            : None
pandas_gbq          : None
pyarrow             : 14.0.1
pyreadstat          : None
pyxlsb              : None
s3fs                : None
scipy               : 1.11.3
sqlalchemy          : None
tables              : 3.9.1
tabulate            : None
xarray              : None
xlrd                : None
zstandard           : None
tzdata              : 2023.3
qtpy                : 2.4.1
pyqt5               : None
</details>

**Comment by tomstarr** [comment]
I'd like to report that I am experiencing this issue which is impacting a project that I'm working on. Previously in Python 3.11.2 (pandas 1.5.3) this was not a problem, but since updating to Python 3.12.1 (pandas 2.1.4) I've hit this issue.

A very simple test case is:
import pandas as pd
print(pd.__version__)
df=pd.DataFrame({"Promote": ['X','NaN','X'," ",None,'X']})
print(df['Promote'])
df.loc[df['Promote'] == 'X', 'promoteperm'] = 'promote'
print(df['promoteperm'])

The output from Python 3.12.1 / pandas 2.1.4 is:
2.1.4
0       X
1     NaN
2       X
3        
4    None
5       X
Name: Promote, dtype: object
0    promote
1        nan
2    promote
3        nan
4        nan
5    promote
Name: promoteperm, dtype: object

In Python 3.11.2 / pandas 1.5.3 the behaviour was as below (which was expected):
1.5.3
0       X
1     NaN
2       X
3        
4    None
5       X
Name: Promote, dtype: object
0    promote
1        NaN
2    promote
3        NaN
4        NaN
5    promote
Name: promoteperm, dtype: object

The impact of this issue is mostly on my lambda joins, many use the dropna function. This function seems to be case sensitive, so it doesn't drop "nan" values.

**Comment by tomstarr** [comment]
After a little further testing, the issue appears to have been introduced in pandas v2.1.2. Perhaps a change was made to address the 'FurtherWarning' which the test case encountered in pandas v2.1.1 (see below)?

2.1.1
0       X
1     NaN
2       X
3        
4    None
5       X
Name: Promote, dtype: object

Warning (from warnings module):
  File "<redacted>", line 5
    df.loc[df['Promote'] == 'X', 'promoteperm'] = 'promote'
FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'promote' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
0    promote
1        NaN
2    promote
3        NaN
4        NaN
5    promote
Name: promoteperm, dtype: object


2.1.2
0       X
1     NaN
2       X
3        
4    None
5       X
Name: Promote, dtype: object
0    promote
1        nan
2    promote
3        nan
4        nan
5    promote
Name: promoteperm, dtype: object

**Comment by DanielHueckstaedt** [comment]
Had the same issue. Could it also trace to the change of pandas 2.1.1 to 2.1.2

Found a work-around.

BEFORE:
>existing_df.loc[bool_filter_pd_series, column_lable] = 'value'
--> ... ... nan
>existing_df.loc[-1, -1].dtype
--> <U3

AFTER:
>existing_df[column_lable] = pd.Series(dtype='str')
>existing_df.loc[bool_filter_pd_series, column_lable] = 'value'
--> ... ... NaN
>existing_df.loc[-1, -1].dtype
--> AttributeError: 'float' object has no attribute 'dtype' (what usualy happens with a real NaN)

**Comment by asishm** [comment]
this is fixed on main (and possibly on the latest version as well)

**Comment by tomstarr** [comment]
Confirming that this issue has been resolved from v2.2.0 onwards.
Tested in v2.2.0, v2.2.1 and v2.2.2
This issue can be closed (IMO).

>>> import pandas as pd
>>> print(pd. __version__)
2.2.0
>>> df=pd.DataFrame({"Promote": ['X','NaN','X'," ",None,'X']})
>>> print(df['Promote'])
0       X
1     NaN
2       X
3
4    None
5       X
Name: Promote, dtype: object
>>> df.loc[df['Promote'] == 'X', 'promoteperm'] = 'promote'
>>> print(df['promoteperm'])
0    promote
1        NaN
2    promote
3        NaN
4        NaN
5    promote
Name: promoteperm, dtype: object
